File: core/src/main/java/io/github/fourlastor/game/demo/round/step/TileSmash.java
package io.github.fourlastor.game.demo.round.step;

import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.scenes.scene2d.Action;
import com.badlogic.gdx.scenes.scene2d.actions.Actions;
import com.badlogic.gdx.scenes.scene2d.actions.SequenceAction;
import com.badlogic.gdx.scenes.scene2d.utils.TextureRegionDrawable;
import dagger.assisted.Assisted;
import dagger.assisted.AssistedFactory;
import dagger.assisted.AssistedInject;
import io.github.fourlastor.game.demo.AttackAnimation;
import io.github.fourlastor.game.demo.state.GameState;
import io.github.fourlastor.game.demo.state.map.Tile;
import io.github.fourlastor.game.demo.state.map.TileType;
import io.github.fourlastor.game.demo.state.unit.Unit;

public class TileSmash extends SimpleStep {

    private final Unit source;
    private final Tile target;
    float moveDuration = 0.05f;
    // Amount to scale the animation by.
    private final Vector3 scale = new Vector3(1f / 16f, 1f, 1f);
    int damage = 2;
    TextureAtlas textureAtlas;

    @AssistedInject
    public TileSmash(@Assisted("source") Unit source, @Assisted("target") Tile target, TextureAtlas textureAtlas) {
        this.source = source;
        this.target = target;
        this.textureAtlas = textureAtlas;
    }

    /**
     * Calculate the angle in degrees between two 2D vectors.
     *
     * @param source The source vector.
     * @param target The target vector.
     * @return The angle in degrees between the vectors.
     */
    public static float calculateAngle(Vector2 source, Vector2 target) {
        return target.cpy().sub(source).angleDeg();
    }

    public Action setupAttackAnimation(float distance, float rotationDegrees) {
        // Base animation goes left-to-right.
        Vector3[] positions = {
            new Vector3(-.25f, 0f, 0f),
            new Vector3(-.125f, 0f, 0f),
            new Vector3(0f, 0f, 2f),
            new Vector3(2f, 0f, 0f),
            new Vector3(8f, 0f, 1f),
            new Vector3(8f, 0f, 0f),
            new Vector3(0f, 0f, 0f),
            new Vector3(0f, 0f, -1f),
            new Vector3(0f, 0f, 0f),
            new Vector3(0f, 0f, -2f),
            new Vector3(-3f, 0f, 16f),
            new Vector3(-3f, 0f, 8f),
            new Vector3(-3f, 0f, 0f),
            new Vector3(-3f, 0f, -8f),
            new Vector3(-3f, 0f, -16f),
        };
        Runnable[] runnables = new Runnable[] {
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            () -> {
                if (target != null) {
                    TextureRegion region = textureAtlas.findRegion("tiles/white");
                    target.actor.setDrawable(new TextureRegionDrawable(region));
                    target.actor.setSize(target.actor.getPrefWidth(), target.actor.getPrefHeight());
                }
            },
            null,
            null,
            null,
            null,
            null,
        };
        Vector3 scale = this.scale.cpy().scl(distance, 1f, 1f);
        return AttackAnimation.makeSequence(source.group, runnables, positions, moveDuration, rotationDegrees, scale);
    }

    @Override
    public void enter(GameState state, Runnable continuation) {

        target.type = TileType.TERRAIN;

        Vector2 originalPosition = new Vector2(source.getActorPosition());
        Vector2 targetPosition = target.getActorPosition();
        // Distance between source and target is used to scale the animation if needed.
        float distance = source.getActorPosition().dst(targetPosition);
        // Angle offset of target from source.
        float rotationDegrees = calculateAngle(originalPosition, targetPosition);
        SequenceAction attackAnimation = Actions.sequence(
                setupAttackAnimation(distance, rotationDegrees),
                Actions.run(() -> source.setActorPosition(originalPosition)),
                Actions.run(continuation));
        source.group.addAction(attackAnimation);
    }

    @AssistedFactory
    public interface Factory {
        TileSmash create(@Assisted("source") Unit source, @Assisted("target") Tile target);
    }
}

File: core/src/main/java/io/github/fourlastor/game/demo/round/step/TileSmash.kt
package io.github.fourlastor.game.demo.round.step

import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.math.Vector3
import com.badlogic.gdx.scenes.scene2d.Action
import com.badlogic.gdx.scenes.scene2d.actions.Actions
import com.badlogic.gdx.scenes.scene2d.utils.TextureRegionDrawable
import dagger.assisted.Assisted
import dagger.assisted.AssistedFactory
import dagger.assisted.AssistedInject
import io.github.fourlastor.game.demo.AttackAnimation.makeSequence
import io.github.fourlastor.game.demo.state.GameState
import io.github.fourlastor.game.demo.state.map.Tile
import io.github.fourlastor.game.demo.state.map.TileType
import io.github.fourlastor.game.demo.state.unit.Unit

class TileSmash @AssistedInject constructor(
    @Assisted("source") private val source: Unit,
    @Assisted("target") private val target: Tile?,
    private val textureAtlas: TextureAtlas
) : SimpleStep() {
    private var moveDuration = 0.05f

    // Amount to scale the animation by.
    private val scale = Vector3(1f / 16f, 1f, 1f)
    private fun setupAttackAnimation(distance: Float, rotationDegrees: Float): Action {
        // Base animation goes left-to-right.
        val positions = arrayOf(
            Vector3(-.25f, 0f, 0f),
            Vector3(-.125f, 0f, 0f),
            Vector3(0f, 0f, 2f),
            Vector3(2f, 0f, 0f),
            Vector3(8f, 0f, 1f),
            Vector3(8f, 0f, 0f),
            Vector3(0f, 0f, 0f),
            Vector3(0f, 0f, -1f),
            Vector3(0f, 0f, 0f),
            Vector3(0f, 0f, -2f),
            Vector3(-3f, 0f, 16f),
            Vector3(-3f, 0f, 8f),
            Vector3(-3f, 0f, 0f),
            Vector3(-3f, 0f, -8f),
            Vector3(-3f, 0f, -16f)
        )
        val runnables = arrayOf(
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Runnable {
                if (target != null) {
                    val region: TextureRegion = textureAtlas.findRegion("tiles/white")
                    target.actor.setDrawable(TextureRegionDrawable(region))
                    target.actor.setSize(target.actor.getPrefWidth(), target.actor.getPrefHeight())
                }
            },
            null,
            null,
            null,
            null,
            null
        )
        val scale = scale.cpy().scl(distance, 1f, 1f)
        return makeSequence(source.group, runnables, positions, moveDuration, rotationDegrees, scale)
    }

    override fun enter(state: GameState, continuation: Runnable) {
        target!!.type = TileType.TERRAIN
        val originalPosition = Vector2(source.actorPosition)
        val targetPosition: Vector2 = target.actorPosition
        // Distance between source and target is used to scale the animation if needed.
        val distance = source.actorPosition.dst(targetPosition)
        // Angle offset of target from source.
        val rotationDegrees = calculateAngle(originalPosition, targetPosition)
        val attackAnimation = Actions.sequence(
            setupAttackAnimation(distance, rotationDegrees),
            Actions.run { source.hex.set(target.hex) }, // (sheerst) Note: model code, likely shouldn't happen here?
            Actions.run { source.actorPosition = originalPosition },
            Actions.run(continuation)
        )
        source.group.addAction(attackAnimation)
    }

    @AssistedFactory
    interface Factory {
        fun create(@Assisted("source") source: Unit, @Assisted("target") target: Tile): TileSmash
    }

    companion object {
        /**
         * Calculate the angle in degrees between two 2D vectors.
         *
         * @param source The source vector.
         * @param target The target vector.
         * @return The angle in degrees between the vectors.
         */
        fun calculateAngle(source: Vector2, target: Vector2): Float {
            return target.cpy().sub(source).angleDeg()
        }
    }
}
